torch/lib/THPP/tensors/generic/THTensor.cpp

```diff
@@ -219,8 +225,269 @@ auto THTensor<real>::free() -> THTensor& {
   return *this;
 }
 
-#define non_const_cast(tensor) const_cast<THTensor&>(dynamic_cast<const THTensor&>(tensor))
-#define non_const_long_cast(tensor) const_cast<THTensor<long>&>(dynamic_cast<const THTensor<long>&>(tensor))
+template<>
+auto THTensor<real>::diag(const Tensor& src, int k) -> THTensor& {
+  THTensor_(diag)(tensor, non_const_cast(src).tensor, k);
+  return *this;
+}
+
+template<>
+auto THTensor<real>::eye(long n, long m) -> THTensor& {
+  THTensor_(eye)(tensor, n, m);
+  return *this;
+}
+
+template<>
+auto THTensor<real>::range(scalar_type xmin, scalar_type xmax,
+                           scalar_type step) -> THTensor& {
+  THTensor_(range)(tensor, xmin, xmax, step);
+  return *this;
+}
+
+// THTensor& THTensor<real>::randperm() {
+//   return *this;
+// }
+
+template<>
+auto THTensor<real>::sort(const Tensor& ri, const Tensor& src,
+             int dimension, int desc) -> THTensor& {
+  throw std::runtime_error("sort is not available yet");
+  THTensor_(sort)(
+    tensor,
+    non_const_long_cast(ri).tensor,
+    non_const_cast(src).tensor,
+    dimension,
+    desc
+  );
+  return *this;
+}
+
+template<>
+auto THTensor<real>::topk(const Tensor& ri, const Tensor& src,
+             long k, int dim, int dir, int sorted) -> THTensor& {
+  throw std::runtime_error("topk is not available yet");
+  THTensor_(topk)(
+    tensor,
+    non_const_long_cast(ri).tensor,
+    non_const_cast(src).tensor,
+    k,
+    dim,
+    dir,
+    sorted
+  );
+  return *this;
+}
+
+template<>
+auto THTensor<real>::tril(const Tensor& src, long k) -> THTensor& {
+  THTensor_(tril)(tensor, non_const_cast(src).tensor, k);
+  return *this;
+}
+
+template<>
+auto THTensor<real>::triu(const Tensor& src, long k) -> THTensor& {
+  THTensor_(triu)(tensor, non_const_cast(src).tensor, k);
+  return *this;
+}
+
+template<>
+auto THTensor<real>::catArray(const std::vector<Tensor*>& inputs_vec,
+                              int dimension) -> THTensor& {
+  int numInputs = inputs_vec.size();
+  tensor_type *inputs[numInputs];
+  for (std::size_t i = 0; i < numInputs; i++)
+    inputs[i] = non_const_cast(*inputs_vec[i]).tensor;
+  THTensor_(catArray)(tensor, inputs, numInputs, dimension);
+  return *this;
+}
+
+template<>
+int THTensor<real>::equal(const Tensor& other) const {
+  return THTensor_(equal)(tensor, non_const_cast(other).tensor);
+}
+
+// Note: the order in _Value and _Tensor is reversed compared to
+// the declarations in master/generic/THDTensorMath.h,
+// so that the first argument is casted onto a byte tensor type
```

Comment:
In `TH`, without the `C++` wrappers, the functions such as `THTensor_(ltTensor)` are declared like [this](https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensorMath.c#L2191) - the "return value", i.e. the first argument, is a `ByteTensor`. The most sane solution would be to define an `ltTensor` method only for `THTensor<unsigned char>`, but we don't want to do that, so after some thought, along with @VirrageS, we have decided that it will be `ta` to call the method with `r` and `tb` as arguments and casting `r` to the appropriate type